examples/hyperparam/search_gpyopt.py

```diff
@@ -0,0 +1,207 @@
+              help="Optimizer algorhitm.")
+@click.option("--seed", type=click.INT, default=97531,
+              help="Seed for the random generator")
+@click.option("--training-experiment-id", type=click.INT, default=-1,
+              help="Maximum number of runs to evaluate. Inherit parent;s experiment if == -1.")
+@click.argument("training_data")
+def run(training_data, max_runs, batch_size, max_p, epochs, metric, gpy_model, gpy_acquisition,
+        initial_design, seed, training_experiment_id):
+    bounds = [
+        {'name': 'lr', 'type': 'continuous', 'domain': (1e-5, 1e-1)},
+        {'name': 'momentum', 'type': 'continuous', 'domain': (0.0, 1.0)},
+    ]
+    # create random file to store run ids of the training tasks
+    tmp = tempfile.mkdtemp()
+    results_path = os.path.join(tmp, "results.txt")
+    tracking_service = mlflow.tracking.get_service()
+
+    def new_eval(nepochs,
+                 experiment_id,
+                 null_train_loss,
+                 null_valid_loss,
+                 null_test_loss,
+                 return_all=False):
+        """
+        Create a new eval function
+
+        :param nepochs: Number of epochs to train the model.
+        :experiment_id: Experiment id for the training run
+        :valid_null_loss: Loss of a null model on the validation dataset
+        :test_null_loss: Loss of a null model on the test dataset.
+        :return_test_loss: Return both validation and test loss if set.
+
+        :return: new eval function.
+        """
+
+        def eval(params):
+            """
+            Train keras model with given parameters by invoking mlflow run.
+
+            Notice we store runUuid and resulting metric in a file. We will later use these to pick
+            the best run and to log the runUuids of the child runs as an artifact. This is a
+            temporary workaround until MLflow offers better mechanism of linking runs together.
+
+            :param params: Parameters to the train_keras script we optimize over:
+                          learning_rate, drop_out_1
+            :return: The metric value evaluated on the validation data.
+            """
+            lr, momentum = params[0]
+            p = mlflow.projects.run(
+                uri=".",
+                entry_point="train",
+                parameters={
+                    "training_data": training_data,
+                    "epochs": str(nepochs),
+                    "learning_rate": str(lr),
+                    "momentum": str(momentum),
+                    "seed": str(seed)},
+                experiment_id=experiment_id,
+                block=False
+            )
+            store = mlflow.tracking._get_store()
+            if p.wait():
+                # cap the loss at the loss of the null model
+                train_loss = min(null_valid_loss,
+                                 store.get_metric(p.run_id, "train_{}".format(metric)).value)
```

Comment:
@tomasatdatabricks We were trying to find a good example that uses the public APIs to get metrics from past runs, and this looks great except that `store.get_metric` is not a public API. Could we use get_run() on the Tracking Service API and parse the metrics out from there? I know it won't look as pretty, but that means we should think about adding a pretty public API rather than having an example with an internal API.

Reply:
Sure, will do, it makes sense to have public apis. We could definitely improve the API here.